lib/AST/ASTStructuralEquivalence.cpp

```diff
@@ -1280,8 +1280,14 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
 
   // Determine whether we've already produced a tentative equivalence for D1.
   Decl *&EquivToD1 = Context.TentativeEquivalences[D1->getCanonicalDecl()];
-  if (EquivToD1)
-    return EquivToD1 == D2->getCanonicalDecl();
+  if (EquivToD1) {
+    if (EquivToD1 == D2->getCanonicalDecl()) // we encountered an already
+                                             // visited (D1, D2) pair
+      return true;
+    else
+      llvm_unreachable("Corrupted AST with multiple redundant definitions in "
```

Comment:
Change definitions -> declarations

Reply:
I am in favor of the change.

Reply:
This assert can fail even in non-wrong cases. It is possible that the same from-Decl is checked against different to-Decl's, like in this case:
```
TEST_F(StructuralEquivalenceTest, CompareSameDeclWithMultiple) {
  auto t = makeNamedDecls(
      "struct A{};struct B{};void foo(A a, A b);",
      "struct A{};struct B{};void foo(A a, B b);",
      Lang_CXX);
  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
}
```

Reply:
Or we could just simply let the control flow onward?
```
if (EquivToD1)
    if (EquivToD1 == D2->getCanonicalDecl()) 
      return true
```
This way we would check (`D1`, `D2'`) as well. And even if we left out something from the redecl chain the eq check could be more reliable.